tests/pep/test_service.py

```diff
@@ -0,0 +1,246 @@
+        self.disco_server = aioxmpp.DiscoServer(self.cc)
+        self.pubsub = aioxmpp.PubSubClient(self.cc, dependencies={
+            aioxmpp.DiscoClient: self.disco_client
+        })
+
+        self.s = pep.PEPClient(self.cc, dependencies={
+            aioxmpp.DiscoClient: self.disco_client,
+            aioxmpp.DiscoServer: self.disco_server,
+            aioxmpp.PubSubClient: self.pubsub
+        })
+
+    def tearDown(self):
+        del self.cc
+        del self.disco_client
+        del self.disco_server
+        del self.pubsub
+        del self.s
+
+    def test_is_service(self):
+        self.assertTrue(issubclass(pep.PEPClient, aioxmpp.service.Service))
+
+    def test_check_for_pep(self):
+        disco_info = disco_xso.InfoQuery()
+        disco_info.identities.append(
+            disco_xso.Identity(type_="pep", category="pubsub")
+        )
+
+        with unittest.mock.patch.object(self.disco_client, "query_info",
+                                        new=CoroutineMock()):
+            self.disco_client.query_info.return_value = disco_info
+            run_coroutine(self.s._check_for_pep())
+            self.disco_client.query_info.assert_called_once_with(
+                TEST_FROM.bare()
+            )
+
+    def test_check_for_pep_failure(self):
+
+        with unittest.mock.patch.object(self.disco_client, "query_info",
+                                        new=CoroutineMock()):
+            self.disco_client.query_info.return_value = disco_xso.InfoQuery()
+
+            with self.assertRaises(RuntimeError):
+                run_coroutine(self.s._check_for_pep())
+
+            self.disco_client.query_info.assert_called_once_with(
+                TEST_FROM.bare()
+            )
+
+    def test_handle_pubsub_publish_is_depsignal_handler(self):
+        self.assertTrue(aioxmpp.service.is_depsignal_handler(
+            aioxmpp.PubSubClient,
+            "on_item_published",
+            self.s._handle_pubsub_publish
+        ))
+
+    def test_publish(self):
+        with contextlib.ExitStack() as stack:
+            stack.enter_context(
+                unittest.mock.patch.object(
+                    self.s,
+                    "_check_for_pep",
+                    CoroutineMock(),
+                )
+            )
+
+            stack.enter_context(
+                unittest.mock.patch.object(
+                    self.pubsub,
+                    "publish",
+                    CoroutineMock(),
+                )
+            )
+
+            run_coroutine(self.s.publish(
+                "urn:example",
+                unittest.mock.sentinel.data,
+                id_="example-id",
+            ))
+
+            self.s._check_for_pep.assert_called_once_with()
+            self.pubsub.publish.assert_called_once_with(
+                None,
+                "urn:example",
+                unittest.mock.sentinel.data,
+                id_="example-id",
+            )
```

Comment:
Yes, I am not entirely sure why I adopted that style. I thought I had a reason and continued with it. Probably your style is cleaner.